### PR TITLE
⚡️

### LINKED ISSUE #64: export node config for ln-sim
Warnet takes in a topology spec and exports all the lightning containers along with their connection info and credentials. we will need to open a port for them to connect to.

```json
 "nodes": [
    {
      "LND": {
        "id": "",
        "address": "",
        "macaroon": "",
        "cert": ""
      }
    },
    {
      "CLN": {
        "id": "",
        "address": "",
        "ca_pem": "",
        "client_pem": "",
        "client_cert": "",
      }
    }
  ],
```

### PR DESCRIPTION
**Pull request by pinheadmz**
WIP: pushing one commit at a time to test on CI

### CODE REVIEW

File: src/warnet/lnnode.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from docker.models.containers import Container
+from warnet.client import get_file_from_container
+from warnet.utils import (
+    exponential_backoff,
+    generate_ipv4_addr
+)
+
+CONTAINER_PREFIX_LN = "ln"
+
+class LNNode:
```
Comment (willcl-ark) [inline src/warnet/lnnode.py:13]: Seems like it might behoove us to make `LNNode` another abstract base class, which can be overridden by an LND node, CLN node or other future node types?

File: src/warnet/lnnode.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import json
+import os
+from docker.models.containers import Container
+from warnet.client import get_file_from_container
+from warnet.utils import (
+    exponential_backoff,
+    generate_ipv4_addr
+)
+
+CONTAINER_PREFIX_LN = "ln"
```
Comment (willcl-ark) [inline src/warnet/lnnode.py:11]:
```suggestion
CONTAINER_PREFIX_LN = "lnd"
```

File: src/scenarios/ln_init.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            mp = self.nodes[3].getrawmempool()
+            for tx in opening_txs:
+                if tx not in mp:
+                    all_set = False
+            if all_set:
+                break
+            sleep(2)
+
+        self.log.info("Confirming channel opens")
+        self.generatetoaddress(self.nodes[3], 6, addr)
```
Comment (willcl-ark) [inline src/scenarios/ln_init.py:59]: I thought we could do 3 confs here, but seems like LND scales num confirmations needed according to channel size, so safest to stick to 6

File: src/warnet/client.py [hunk 4]
```diff
@@ -6,28 +6,29 @@
...
 
-
-def get_bitcoin_debug_log(network: str, index: int) -> str:
-    warnet = Warnet.from_network(network)
-    tank = warnet.tanks[index]
-    subdir = "/" if tank.bitcoin_network == "main" else f"{tank.bitcoin_network}/"
-    data, stat = tank.container.get_archive(f"/home/bitcoin/.bitcoin/{subdir}debug.log")
-    out = ""
+def get_file_from_container(container, path):
+    data, stat = container.get_archive(path)
```
Comment (willcl-ark) [inline src/warnet/client.py]: I think we'll also want to add a container abstract base class method for `get_file` here so that we can keep it portable between docker and k8s

File: src/warnet/client.py [hunk 5]
```diff
@@ -94,10 +92,9 @@ def thread_stop():
     return True
 
 
-def compose_down(network="warnet") -> bool:
+def compose_down(warnet) -> bool:
```
Comment (willcl-ark) [inline src/warnet/client.py]: unrelated change? :P

File: src/warnet/lnnode.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            "image": "lightninglabs/lnd:v0.17.0-beta",
+            "command": " ".join(args),
+            "networks": {
+                self.tank.docker_network: {
+                    "ipv4_address": f"{self.ipv4}",
+                }
+            }
+        }
+
+    @exponential_backoff(max_retries=20, max_delay=300)
```
Comment (willcl-ark) [inline src/warnet/lnnode.py:73]: 20 retries? Do these take a long time to come up? Perhaps we can wait until they're online before issuing RPC commands?

File: src/warnet/warnet.py [hunk 7]
```diff
@@ -42,10 +44,17 @@ def __init__(self, config_dir):
 
 
     def __str__(self) -> str:
-        template = "\t%-8.8s%-25.24s%-25.24s%-25.24s%-18.18s\n"
-        tanks_str = template % ("Index", "Version", "Conf", "Netem", "IPv4")
+        # TODO: bitcoin_conf and tc_netem can be added back in to this table
```
Comment (willcl-ark) [inline src/warnet/warnet.py:36]: Could use tabulate to do this for us? No need to re-invent the wheel, and I think adding a dep for it is OK?

```sh
pip install tabulate
```

```python
from tabulate import tabulate
    
    def __str__(self) -> str:
        headers = ["Index", "Version", "Conf", "Netem", "IPv4", "LN", "LN IPV4"]
        tanks_data = []
        for tank in self.tanks:
        	tanks_data.append([tank.index, tank.version, tank.conf, tank.netem, tank.ipv4, tank.lnnode.impl if tank.lnnode is not None else None, tank.lnnode.ipv4 if tank.lnnode is not None else None])

        return tabulate(tanks_data, headers=headers, tablefmt="plain")
```

**Review by willcl-ark — CHANGES_REQUESTED** [review]
Did a light code review, but didn't try running anything.

Looks like it needs quite a hefty rebase on current master, but otherwise pretty nice!

Left a few comments on various files.

File: src/warnet/lnnode.py (hunk 1 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:13]: Yeah thats the plan. Just haven;t gotten to CLN yet

File: src/warnet/lnnode.py (hunk 2 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:11]: no because the container could be a CLN node

File: src/warnet/client.py (hunk 5 above)
Comment (pinheadmz) [reply src/warnet/client.py]: 🤔 hmm what the heck was I thinking

File: src/warnet/warnet.py (hunk 7 above)
Comment (pinheadmz) [reply src/warnet/warnet.py:36]: sure but maybe in a separate clean up PR since there are a few tables through our RPC.

Comment (willcl-ark) [reply src/warnet/warnet.py:36]: I may draft up just such a tabulation PR to see what it would look like for the others...

**Comment by pinheadmz** [comment]
Rebased on main!
TODO:
- [x] test again locally with actual sim-ln
- [x] cover `warcli export` command in test (followup)
- [ ] add CLN (followup)
- [x] docs

**Comment by willcl-ark** [comment]
Wow nice one. Will review

**Comment by pinheadmz** [comment]
added a commit that fixes `warcli network export` and covers with test. I also ran sim-ln successfully using the exported data from warnet! last step here is to write docs and hand the branch over to sim-ln team to see if the docs and code is sufficient.

**Comment by pinheadmz** [comment]
added docs!

File: src/warnet/lnnode.py (hunk 2 above)
Comment (willcl-ark) [reply src/warnet/lnnode.py:11]: Sure, so we'll get this working first then abstract it later?

File: src/warnet/lnnode.py (hunk 6 above)
Comment (willcl-ark) [reply src/warnet/lnnode.py:73]: Why is this set so high?

File: src/warnet/lnnode.py (hunk 2 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:11]: yeah

File: src/warnet/lnnode.py (hunk 6 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:73]: I think I was effectively using this as a health check bc lnd can take so long to be responsive after startup. Might not be necessary

File: src/warnet/lnnode.py (hunk 2 above)
Comment (willcl-ark) [reply src/warnet/lnnode.py:11]: And we won't give these a cutsie prefix like `bolt` or `spark` :'(

File: src/warnet/lnnode.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+class LNNode:
+    def __init__(self, warnet, tank, impl):
+        self.warnet = warnet
+        self.tank = tank
+        assert impl == "lnd"
+        self.impl = impl
+        self._container = None
+        self.container_name = f"{self.tank.docker_network}_{CONTAINER_PREFIX_LN}_{self.tank.suffix}"
+        self.ipv4 = generate_ipv4_addr(self.warnet.subnet)
```
Comment (willcl-ark) [inline src/warnet/lnnode.py:21]: Why don't we reuse the `tank` ipv4 address here? It will have it's own port, no?

File: src/warnet/lnnode.py (hunk 2 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:11]: lets just go back to using Winston Churchill quotes for everything

File: src/warnet/lnnode.py (hunk 8 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:21]: this is the ip address of the ln node, it lives in a different container from the tank

Comment (pinheadmz) [reply src/warnet/lnnode.py:21]: oh i see what you mean, same IP different port. hmmmmmm

Comment (pinheadmz) [reply src/warnet/lnnode.py:21]: My assistant says containers should all have their own IPs even if they use different ports

File: docs/lightning.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+## Running the Lightning network
+
+When warnet is started with `warcli network start` the bitcoin containers will
+be started first followed by the lightning node containers. It may require a few
+automatic restarts before the lightning nodes start up and connect to their
+corresponding bitcoin nodes. Use `warcli network status` to monitor container status
+and wait for all containers to be `running`.
+
+To create the lightning channels specified in the graph file, run the included
```
Comment (willcl-ark) [inline docs/lightning.md:45]: Feels slightly odd to have this as a scenario. Not a big deal, but perhaps we could in the future move this to it's own subcommand, `warcli simln init` or similar?

File: docs/lightning.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+```
+$ warcli network export
+/Users/bitcoin-dev-project/.warnet/warnet/warnet/simln
+
+$ ls /Users/bitcoin-dev-project/.warnet/warnet/warnet/simln
+sim.json                         warnet_ln_000000_tls.cert        warnet_ln_000001_tls.cert        warnet_ln_000002_tls.cert
+warnet_ln_000000_admin.macaroon  warnet_ln_000001_admin.macaroon  warnet_ln_000002_admin.macaroon
+
+$ sim-cli /Users/bitcoin-dev-project/.warnet/warnet/warnet/simln/sim.json
```
Comment (willcl-ark) [inline docs/lightning.md]:
```suggestion
$ sim-cli --data-dir /Users/bitcoin-dev-project/.warnet/warnet/warnet/simln/sim.json 
```

File: src/interfaces/docker_interface.py [hunk 11]
```diff
@@ -376,11 +392,16 @@ def tank_from_deployment(self, service, warnet):
...
         if "image" in service:
-            t.version = service["image"].split(":")[1]
+            tank.version = service["image"].split(":")[1]
         else:
-            t.version = f"{service['build']['args']['REPO']}#{service['build']['args']['BRANCH']}"
-        return t
-
+            tank.version = f"{service['build']['args']['REPO']}#{service['build']['args']['BRANCH']}"
+
+        labels = service.get("labels", {})
```
Comment (willcl-ark) [inline src/interfaces/docker_interface.py:402]: The labels are a nice trick, but could mess up some prometheus/grafana stuff if we added any of that onto the ln nodes.

Again, I don't think this is something we should worry much about _now_ though.

File: src/interfaces/interfaces.py [hunk 12]
```diff
@@ -60,6 +60,13 @@ def get_bitcoin_cli(self, container_name: str, method: str, params=None):
         """
         raise NotImplementedError("This method should be overridden by child class")
 
+    @abstractmethod
+    def get_file_from_container(self, container_name: str, path: str):
```
Comment (willcl-ark) [inline src/interfaces/interfaces.py]:
```suggestion
    def get_file(self, container_name: str, file_path: str):
```

File: src/interfaces/docker_interface.py [hunk 13]
```diff
@@ -130,8 +131,19 @@ def get_bitcoin_cli(self, tank: Tank, method: str, params=None):
             cmd = f"bitcoin-cli -regtest -rpcuser={tank.rpc_user} -rpcport={tank.rpc_port} -rpcpassword={tank.rpc_password} {method}"
         return self.exec_run(tank.container_name, cmd, user="bitcoin")
 
+    def get_file_from_container(self, container_name, path):
```
Comment (willcl-ark) [inline src/interfaces/docker_interface.py]:
```suggestion
    def get_file(self, container_name, file_path):
```

slight preference for breveity

File: src/warnet/cli/main.py [hunk 14]
```diff
@@ -146,7 +146,7 @@ def grep_logs(pattern, network):
     except Exception as e:
         print(
                 f"Error fetching combined log: {e}"
-        )
+            )
```
Comment (willcl-ark) [inline src/warnet/cli/main.py]: nit: think this was correct before?

File: src/warnet/lnnode.py (hunk 8 above)
Comment (willcl-ark) [reply src/warnet/lnnode.py:21]: Yes they very much should!

File: src/warnet/lnnode.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+            "--debuglevel=debug",
+            "--accept-keysend",
+            "--bitcoin.active",
+            "--bitcoin.regtest",
+            "--bitcoin.node=bitcoind",
+            f"--bitcoind.rpcuser={self.tank.rpc_user}",
+            f"--bitcoind.rpcpass={self.tank.rpc_password}",
+            f"--bitcoind.rpchost={self.tank.ipv4}:{self.tank.rpc_port}",
+            f"--bitcoind.zmqpubrawblock=tcp://{self.tank.ipv4}:28332",
+            f"--bitcoind.zmqpubrawtx=tcp://{self.tank.ipv4}:28333",
```
Comment (willcl-ark) [inline src/warnet/lnnode.py]: Use properties instead of magic constants in `docker_interface.py`, here and just below in `rpclisten` param?

Although... I suppose we don't _set_ them explicitly for bitcoin nodes... hmmm

File: src/warnet/lnnode.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+            f"--externalip={self.ipv4}",
+            f"--rpclisten=0.0.0.0:10009",
+            f"--alias={self.container_name}"
+        ]
+        services[self.container_name] = {
+            "container_name": self.container_name,
+            "image": "lightninglabs/lnd:v0.17.0-beta",
```
Comment (willcl-ark) [inline src/warnet/lnnode.py:54]: Just curious, is this version planned to be configurable in the future?

File: src/warnet/lnnode.py (hunk 6 above)
Comment (willcl-ark) [reply src/warnet/lnnode.py:73]: I see. I can't think of much better way than doing this from outside without calling into `docker ps -a` and doing filtering on the text output right now, which would be super ugly...

File: src/interfaces/docker_interface.py [hunk 17]
```diff
@@ -275,6 +278,8 @@ def default_config_args(self, tank):
         defaults += f" -rpcuser={tank.rpc_user}"
         defaults += f" -rpcpassword={tank.rpc_password}"
         defaults += f" -rpcport={tank.rpc_port}"
+        defaults +=  " -zmqpubrawblock=tcp://0.0.0.0:28332"
```
Comment (willcl-ark) [inline src/interfaces/docker_interface.py]: There are those magic numbers!

File: src/warnet/warnet.py (hunk 7 above)
Comment (willcl-ark) [reply src/warnet/warnet.py:36]: Leaving as unresolved just so I remember to do it in the future.

**Review by willcl-ark — APPROVED** [review]
ACK 8e4be123650a2b861bcdb4ea287b6090e6a78a3a

left more comments, nits and other bits which can be addressed here or in a followup. Happy to merge this as-is.

I tested the changes and found the integration worked. Followed the docs and `Sim-ln` connected perfectly to the `Warnet`. Only issue there was that there is no activity on the simln side in the example json, so it didn't make any payments.

```log
will@ubuntu in ~/src/sim-ln on  main via 🦀 v1.72.0
₿ sim-cli -d /home/will/.config/warnet/warnet/simln
2023-11-20T15:52:17.554Z INFO  [sim_cli] Connected to warnet_ln_000000 - Node ID: 0358785c8359f01ae176c03099fd3529d226e2fc0c24fe8f4de7d2211ab4cfdead.
2023-11-20T15:52:17.603Z INFO  [sim_cli] Connected to warnet_ln_000001 - Node ID: 031406195be86cea35ed9544ad518c5fdf1106898aee7035704a127268365f8338.
2023-11-20T15:52:17.651Z INFO  [sim_cli] Connected to warnet_ln_000002 - Node ID: 0287e195011604427453c46330dc2106b478728d933ac59f5ae658a38b07991130.
2023-11-20T15:52:17.651Z INFO  [sim_lib] Running the simulation forever.
2023-11-20T15:52:17.656Z INFO  [sim_lib] Simulation is running on regtest.
2023-11-20T15:52:17.656Z INFO  [sim_lib] Simulating 0 activity on 3 nodes.
2023-11-20T15:52:17.656Z INFO  [sim_lib] Summary of results will be reported every 60s.
2023-11-20T15:52:17.659Z INFO  [sim_lib] Generating random activity with multiplier: 2, average payment amount: 3800000.
2023-11-20T15:52:17.659Z INFO  [sim_lib] Created network generator: network graph view with: 3 channels.
2023-11-20T15:52:17.659Z INFO  [sim_lib] Started random activity producer for warnet_ln_000002(0287e1...991130): activity generator for capacity: 50000000 with multiplier 2: 26.31578947368421 payments per month (0.03654970760233918 per hour).
2023-11-20T15:52:17.659Z INFO  [sim_lib] Started random activity producer for warnet_ln_000001(031406...5f8338): activity generator for capacity: 100000000 with multiplier 2: 52.63157894736842 payments per month (0.07309941520467836 per hour).
2023-11-20T15:52:17.659Z INFO  [sim_lib] Started random activity producer for warnet_ln_000000(035878...cfdead): activity generator for capacity: 50000000 with multiplier 2: 26.31578947368421 payments per month (0.03654970760233918 per hour).
2023-11-20T15:53:17.658Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T15:54:17.660Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T15:55:17.661Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T15:56:17.661Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T15:57:17.663Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T15:58:17.664Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T15:59:17.665Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T16:00:17.667Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T16:01:17.668Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T16:02:17.670Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
2023-11-20T16:03:17.671Z INFO  [sim_lib] Processed 0 payments sending 0 msat total with NaN% success rate.
```

File: docs/lightning.md (hunk 10 above)
Comment (pinheadmz) [reply docs/lightning.md]: heh good catch, they JUST changed this. I'll update.

https://github.com/bitcoin-dev-project/sim-ln/pull/132/files

File: src/warnet/lnnode.py (hunk 16 above)
Comment (pinheadmz) [reply src/warnet/lnnode.py:54]: yeah we can expand this in the future. lightninglabs doesnt maintain a "latest" tag and all their releases are tagged beta as far as I can tell, so.

File: src/interfaces/docker_interface.py (hunk 17 above)
Comment (pinheadmz) [reply src/interfaces/docker_interface.py]: ok Ill move em

**Comment by willcl-ark** [comment]
🥳

File: src/warnet/warnet.py (hunk 7 above)
Comment (willcl-ark) [reply src/warnet/warnet.py:36]: Opened https://github.com/bitcoin-dev-project/warnet/pull/201
